File: trabajo 17/lista enlazada doblemente circular.java
import java.util.InputMismatchException;
import java.util.Scanner;

/**
 * Clase principal que contiene la implementación de la Lista Doblemente Enlazada Circular
 * y el menú de interacción en la consola.
 */
public class DoublyCircularLinkedListApp {

    /**
     * Clase para representar un nodo en la lista doblemente enlazada.
     * Contiene el dato, y referencias al nodo siguiente y anterior.
     */
    class Node {
        int data;
        Node next; // Referencia al siguiente nodo
        Node prev; // Referencia al nodo anterior

        public Node(int data) {
            this.data = data;
            this.next = null;
            this.prev = null;
        }
    }

    private Node head; // Referencia al primer nodo (cabeza)

    public DoublyCircularLinkedListApp() {
        this.head = null;
    }

    /**
     * Verifica si la lista está vacía.
     * @return true si la cabeza es null.
     */
    public boolean isEmpty() {
        return head == null;
    }

    /**
     * Encuentra y devuelve el último nodo de la lista.
     * En una lista doblemente enlazada circular no vacía, el último nodo es siempre head.prev.
     * @return El último nodo o null si la lista está vacía.
     */
    private Node getLastNode() {
        if (isEmpty()) {
            return null;
        }
        return head.prev;
    }

    // --- Funciones de Inserción ---

    /**
     * Inserta un nodo al principio (antes del head).
     * @param item El valor a insertar.
     */
    public void beginInsert(int item) {
        Node newNode = new Node(item);

        if (isEmpty()) {
            head = newNode;
            // En una lista circular de un solo nodo, apunta a sí mismo
            head.next = head;
            head.prev = head;
        } else {
            Node last = getLastNode();

            // 1. Conexiones del nuevo nodo
            newNode.next = head;
            newNode.prev = last;

            // 2. Reconexiones de los nodos existentes
            head.prev = newNode;
            last.next = newNode;

            // 3. Mover la cabeza al nuevo nodo
            head = newNode;
        }

        System.out.println("\nNodo " + item + " insertado al principio (nuevo head).");
    }

    /**
     * Inserta un nodo al final.
     * @param item El valor a insertar.
     */
    public void lastInsert(int item) {
        Node newNode = new Node(item);

        if (isEmpty()) {
            // Reutilizamos la lógica del primer elemento
            head = newNode;
            head.next = head;
            head.prev = head;
        } else {
            Node last = getLastNode();

            // 1. Conexiones del nuevo nodo
            newNode.next = head;
            newNode.prev = last;

            // 2. Reconexiones de los nodos existentes
            last.next = newNode;
            head.prev = newNode;
        }

        System.out.println("\nNodo " + item + " insertado al final.");
    }

    /**
     * Inserta un nodo DESPUÉS del nodo en la posición (índice 0-basado) dada 'loc'.
     * @param item El valor a insertar.
     * @param loc La posición del nodo ANTERIOR al nuevo.
     */
    public void randomInsert(int item, int loc) {
        if (loc < 0) {
            System.out.println("\nLa ubicación debe ser un número positivo o cero.");
            return;
        }

        if (isEmpty()) {
            System.out.println("\nLa lista está vacía. Use insertar al principio/final.");
            return;
        }
        
        // 1. Encontrar el nodo en la posición 'loc' (el nodo ANTERIOR al nuevo)
        Node prevNode = head;
        int count = 0;

        // Caso especial: si loc es el índice del último nodo, insertamos al final.
        // Pero primero recorremos hasta 'loc'.
        while (count < loc) {
            prevNode = prevNode.next;
            count++;
            
            // Si damos la vuelta al círculo antes de alcanzar 'loc'
            if (prevNode == head && count <= loc) {
                System.out.println("\nNo se puede insertar. La posición está fuera de rango.");
                return;
            }
        }
        
        // Si 'prevNode.next' es 'head' y 'count' es la posición 'loc', estamos en el último nodo.
        // En este caso, randomInsert(item, N-1) donde N es el tamaño, es equivalente a lastInsert(item).
        if (prevNode.next == head && count == loc) {
            lastInsert(item); // Reutilizar la función de inserción al final
            return;
        }
        
        // Si la lista no es de un solo elemento y el índice 'loc' está fuera de rango
        if (prevNode == head && loc > 0) {
            System.out.println("\nNo se puede insertar. La posición está fuera de rango.");
            return;
        }
        
        // Caso intermedio:
        // Nodos: A (prevNode), B (newNode), C (prevNode.next)
        Node newNode = new Node(item);
        Node nextNode = prevNode.next;

        // Conexiones: B <-> A
        newNode.prev = prevNode;
        prevNode.next = newNode;

        // Conexiones: B <-> C
        newNode.next = nextNode;
        nextNode.prev = newNode;

        System.out.println("\nNodo " + item + " insertado después de la posición " + loc + ".");
    }

    // --- Funciones de Eliminación ---

    /**
     * Elimina el nodo al principio (el head actual).
     */
    public void beginDelete() {
        if (isEmpty()) {
            System.out.println("\nLa lista está vacía");
            return;
        }

        int deletedValue = head.data;

        if (head.next == head) {
            // Caso de un solo nodo
            head = null;
        } else {
            Node last = getLastNode();
            
            // 1. Mover head al siguiente
            head = head.next;

            // 2. Actualizar la conexión circular (last <-> new head)
            head.prev = last;
            last.next = head;
        }

        System.out.println("\nNodo con valor " + deletedValue + " eliminado desde el principio (head) ...");
    }

    /**
     * Elimina el último nodo de la lista.
     */
    public void lastDelete() {
        if (isEmpty()) {
            System.out.println("\nLa lista está vacía");
            return;
        }

        if (head.next == head) {
            // Caso de un solo nodo
            int deletedValue = head.data;
            head = null;
            System.out.println("\nNodo con valor " + deletedValue + " eliminado del final ...");
            return;
        }

        Node last = getLastNode(); // Nodo a eliminar
        int deletedValue = last.data;
        Node newLast = last.prev; // El nuevo último

        // 1. El nuevo último apunta al head
        newLast.next = head;

        // 2. El head apunta su 'prev' al nuevo último
        head.prev = newLast;
        
        // Opcional: Desvincular el nodo eliminado (Java lo maneja con GC, pero es buena práctica en teoría de listas)
        last.next = null;
        last.prev = null;

        System.out.println("\nNodo con valor " + deletedValue + " eliminado del final ...");
    }

    /**
     * Elimina el nodo en una posición (índice 0-basado) dada.
     * @param loc La posición (índice 0-basado) a eliminar.
     */
    public void randomDelete(int loc) {
        if (isEmpty()) {
            System.out.println("\nLista vacía. No se puede eliminar.");
            return;
        }

        if (loc < 0) {
            System.out.println("\nLa ubicación debe ser un número positivo o cero.");
            return;
        }

        if (loc == 0) {
            // Delegar en la función de eliminación del principio
            beginDelete();
            return;
        }
        
        Node ptrToDelete = head;
        int count = 0;
        boolean found = false;

        // Avanza hasta la posición 'loc' (el nodo a eliminar)
        do {
            if (count == loc) {
                found = true;
                break;
            }
            
            ptrToDelete = ptrToDelete.next;
            count++;

        } while (ptrToDelete != head);

        if (!found) {
            System.out.println("\nNo se puede eliminar. La posición está fuera de rango.");
            return;
        }
            
        // Si es el último nodo (el nodo anterior al head), delegamos en lastDelete
        if (ptrToDelete.next == head) {
            lastDelete();
            return;
        }

        // Caso intermedio: Nodos A (prev), B (ptrToDelete), C (next)
        int deletedValue = ptrToDelete.data;
        Node prevNode = ptrToDelete.prev;
        Node nextNode = ptrToDelete.next;

        // 1. Reconexión A a C
        prevNode.next = nextNode;

        // 2. Reconexión C a A
        nextNode.prev = prevNode;
        
        // Opcional: Desvincular el nodo eliminado
        ptrToDelete.next = null;
        ptrToDelete.prev = null;

        System.out.println("\nNodo con valor " + deletedValue + " en la posición " + loc + " (índice 0-basado) eliminado.");
    }

    // --- Otras Funciones ---

    /**
     * Busca un elemento y reporta su posición.
     * @param item El valor a buscar.
     */
    public void search(int item) {
        if (isEmpty()) {
            System.out.println("\nLista vacía");
            return;
        }

        Node ptr = head;
        int index = 0;
        boolean found = false;
        StringBuilder positions = new StringBuilder();

        // Recorrido circular
        do {
            if (ptr.data == item) {
                positions.append(index).append(" ");
                found = true;
            }
            
            ptr = ptr.next;
            index++;
            
        } while (ptr != head);

        if (found) {
            System.out.println("Elemento " + item + " encontrado en la(s) ubicación(es) (índice 0-basado): " + positions.toString().trim());
        } else {
            System.out.println("Elemento " + item + " no encontrado");
        }
    }

    /**
     * Muestra todos los elementos de la lista desde head.
     */
    public void display() {
        if (isEmpty()) {
            System.out.println("Lista vacía. Nada que imprimir.");
        } else {
            System.out.println("\nImprimiendo valores (desde HEAD): ");
            Node ptr = head;
            StringBuilder output = new StringBuilder();
            
            // Recorrido circular
            do {
                output.append(ptr.data);
                
                // Mostrar la conexión
                if (ptr.next != head) {
                    output.append(" <-> ");
                } else {
                    // Mostrar la conexión circular
                    output.append(" <-> (Circular) <-> ");
                }

                ptr = ptr.next;
                
            } while (ptr != head);

            // Imprimir el head de nuevo para cerrar el círculo visualmente
            output.append("[... ").append(head.data).append(" (HEAD) ...]");
            System.out.println(output.toString());
        }
    }

    // Función principal para el menú de interacción
    public static void main(String[] args) {
        DoublyCircularLinkedListApp listInstance = new DoublyCircularLinkedListApp();
        Scanner scanner = new Scanner(System.in);
        int choice = 0;
        
        System.out.println("Iniciando Lista Doblemente Enlazada Circular...");
        
        while (choice != 9) {
            try {
                System.out.println("\n\n Menú principal (Lista Doblemente Enlazada Circular) ");
                System.out.println("\nElige una opción de la siguiente lista ...\n");
                System.out.println("------------------------------------------");
                
                System.out.println("1. Insertar al principio (nuevo head)");
                System.out.println("2. Insertar al final");
                System.out.println("3. Insertar en posición (índice 0-basado, *después* del nodo en el índice)");
                System.out.println("4. Eliminar del principio (head)");
                System.out.println("5. Eliminar desde el final");
                System.out.println("6. Eliminar nodo por posición (índice 0-basado)");
                System.out.println("7. Buscar un elemento");
                System.out.println("8. Mostrar (recorrido circular)");
                System.out.println("9. Salir");
                System.out.println("------------------------------------------");
                
                System.out.print("\nIngrese su opción? ");
                choice = scanner.nextInt();

                if (choice == 9) {
                    System.out.println("Saliendo del programa.");
                    break;
                }

                if (choice < 1 || choice > 8) {
                    System.out.println("Introduzca una opción válida (1-9).");
                    continue;
                }

                Integer item = null;
                Integer loc = null;

                if (choice == 1 || choice == 2 || choice == 3 || choice == 7) { 
                    // Requieren un valor (item)
                    System.out.print("\nIngrese el valor (entero) del elemento: ");
                    item = scanner.nextInt();
                }

                if (choice == 3 || choice == 6) { 
                    // Requieren una posición (loc)
                    System.out.print("Introduce la ubicación (índice 0-basado): ");
                    loc = scanner.nextInt();
                }

                switch (choice) {
                    case 1:
                        listInstance.beginInsert(item);
                        break;
                    case 2:
                        listInstance.lastInsert(item);
                        break;
                    case 3:
                        listInstance.randomInsert(item, loc);
                        break;
                    case 4:
                        listInstance.beginDelete();
                        break;
                    case 5:
                        listInstance.lastDelete();
                        break;
                    case 6:
                        listInstance.randomDelete(loc);
                        break;
                    case 7:
                        listInstance.search(item);
                        break;
                    case 8:
                        listInstance.display();
                        break;
                    default:
                        // No debería llegar aquí debido a la verificación previa
                        System.out.println("Opción inválida.");
                        break;
                }

            } catch (InputMismatchException e) {
                // Maneja cuando el usuario ingresa texto en lugar de un número
                System.out.println("\nEntrada inválida. Por favor, ingrese un número entero para la opción/valor/ubicación.");
                scanner.next(); // Consume la entrada incorrecta para evitar un bucle infinito
            } catch (Exception e) {
                System.out.println("\nOcurrió un error inesperado: " + e.getMessage());
                // En un caso real, aquí se imprimiría e.printStackTrace();
            }
        }
        scanner.close(); // Cerrar el scanner al salir del bucle principal
    }
}

File: trabajo 17/lista enlazada doblemente circular.py
import sys

# Definición de la clase para cada nodo
class Node:
    """
    Clase para representar un nodo en la lista doblemente enlazada.
    Contiene el dato, y referencias al nodo siguiente y anterior.
    """
    def __init__(self, data):
        self.data = data
        self.next = None # Referencia al siguiente nodo
        self.prev = None # Referencia al nodo anterior

# Clase que contiene la lógica de la Lista Doblemente Enlazada Circular
class DoublyCircularLinkedList:
    def __init__(self):
        self.head = None # Referencia al primer nodo (cabeza)

    def _get_last_node(self):
        """
        Encuentra y devuelve el último nodo de la lista.
        En una lista doblemente enlazada circular no vacía, el último nodo
        es siempre head.prev.
        """
        if self.is_empty():
            return None
        return self.head.prev

    def is_empty(self):
        """
        Verifica si la lista está vacía.
        """
        return self.head is None

    # --- Funciones de Inserción ---

    def begin_insert(self, item):
        """
        Inserta un nodo al principio (antes del head).
        """
        new_node = Node(item)

        if self.is_empty():
            self.head = new_node
            # En una lista circular de un solo nodo, apunta a sí mismo
            self.head.next = self.head
            self.head.prev = self.head
        else:
            last = self._get_last_node()

            # 1. Conexiones del nuevo nodo
            new_node.next = self.head
            new_node.prev = last

            # 2. Reconexiones de los nodos existentes
            self.head.prev = new_node
            last.next = new_node

            # 3. Mover la cabeza al nuevo nodo
            self.head = new_node
            
        print(f"\nNodo {item} insertado al principio (nuevo head).")

    def last_insert(self, item):
        """
        Inserta un nodo al final.
        """
        new_node = Node(item)

        if self.is_empty():
            # Reutilizamos la lógica del primer elemento
            self.head = new_node
            self.head.next = self.head
            self.head.prev = self.head
        else:
            last = self._get_last_node()

            # 1. Conexiones del nuevo nodo
            new_node.next = self.head
            new_node.prev = last

            # 2. Reconexiones de los nodos existentes
            last.next = new_node
            self.head.prev = new_node
            
        print(f"\nNodo {item} insertado al final.")

    def random_insert(self, item, loc):
        """
        Inserta un nodo DESPUÉS del nodo en la posición (índice 0-basado) dada 'loc'.
        """
        if loc < 0:
            print("\nLa ubicación debe ser un número positivo o cero.")
            return

        if self.is_empty():
            print("\nLa lista está vacía. Use insertar al principio/final.")
            return

        # 1. Encontrar el nodo en la posición 'loc' (el nodo ANTERIOR al nuevo)
        prev_node = self.head
        count = 0
        
        while count < loc:
            prev_node = prev_node.next
            count += 1
            if prev_node == self.head and count <= loc:
                # Si damos la vuelta al círculo antes de alcanzar 'loc'
                print("\nNo se puede insertar. La posición está fuera de rango.")
                return

        # Si 'prev_node' es el último nodo y el índice coincide con 'loc', insertamos al final.
        # En una lista circular, esto se maneja con la lógica general de inserción,
        # pero verificamos si 'prev_node' ya es el último
        if prev_node.next == self.head and count == loc:
            self.last_insert(item) # Reutilizar la función de inserción al final
            return

        # 2. 'prev_node' es el nodo *anterior* donde queremos insertar.
        # Nodos: A (prev_node), B (newNode), C (prev_node.next)
        new_node = Node(item)
        next_node = prev_node.next

        # Conexiones: B <-> A
        new_node.prev = prev_node
        prev_node.next = new_node

        # Conexiones: B <-> C
        new_node.next = next_node
        next_node.prev = new_node

        print(f"\nNodo {item} insertado después de la posición {loc}.")

    # --- Funciones de Eliminación ---

    def begin_delete(self):
        """
        Elimina el nodo al principio (el head actual).
        """
        if self.is_empty():
            print("\nLa lista está vacía")
            return

        deleted_value = self.head.data

        if self.head.next == self.head:
            # Caso de un solo nodo
            self.head = None
        else:
            last = self._get_last_node()
            old_head = self.head
            
            # 1. Mover head al siguiente
            self.head = self.head.next

            # 2. Actualizar la conexión circular (last <-> new head)
            self.head.prev = last
            last.next = self.head

            # Nota: En Python, no necesitamos 'delete old_head' explícito,
            # el recolector de basura liberará la memoria cuando no haya referencias.

        print(f"\nNodo con valor {deleted_value} eliminado desde el principio (head) ...")

    def last_delete(self):
        """
        Elimina el último nodo de la lista.
        """
        if self.is_empty():
            print("\nLa lista está vacía")
            return

        if self.head.next == self.head:
            # Caso de un solo nodo
            deleted_value = self.head.data
            self.head = None
            print(f"\nNodo con valor {deleted_value} eliminado del final ...")
            return

        last = self._get_last_node() # Nodo a eliminar
        deleted_value = last.data
        new_last = last.prev # El nuevo último

        # 1. El nuevo último apunta al head
        new_last.next = self.head

        # 2. El head apunta su 'prev' al nuevo último
        self.head.prev = new_last
        
        # Nota: La memoria del nodo 'last' será liberada por el GC de Python.

        print(f"\nNodo con valor {deleted_value} eliminado del final ...")

    def random_delete(self, loc):
        """
        Elimina el nodo en una posición (índice 0-basado) dada.
        """
        if self.is_empty():
            print("\nLista vacía. No se puede eliminar.")
            return

        if loc < 0:
            print("\nLa ubicación debe ser un número positivo.")
            return

        if loc == 0:
            # Delegar en la función de eliminación del principio
            self.begin_delete()
            return
            
        ptr_to_delete = self.head
        count = 0
        found = False

        # Avanza hasta la posición 'loc' (el nodo a eliminar)
        # Usamos un bucle while True para emular el do-while hasta que demos la vuelta
        while True:
            if count == loc:
                found = True
                break
            
            ptr_to_delete = ptr_to_delete.next
            count += 1

            if ptr_to_delete == self.head:
                # Dimos la vuelta y no encontramos el índice
                break

        if not found:
            print("\nNo se puede eliminar. La posición está fuera de rango.")
            return
            
        # Si es el último nodo (el nodo anterior al head), delegamos en last_delete
        if ptr_to_delete.next == self.head:
            self.last_delete()
            return

        # Caso intermedio: Nodos A (prev), B (ptr_to_delete), C (next)
        deleted_value = ptr_to_delete.data
        prev_node = ptr_to_delete.prev
        next_node = ptr_to_delete.next

        # 1. Reconexión A a C
        prev_node.next = next_node

        # 2. Reconexión C a A
        next_node.prev = prev_node
        
        # Nota: La memoria de ptr_to_delete será liberada por el GC de Python.

        print(f"\nNodo con valor {deleted_value} en la posición {loc} (índice 0-basado) eliminado.")

    # --- Otras Funciones ---

    def search(self, item):
        """
        Busca un elemento y reporta su posición.
        """
        if self.is_empty():
            print("\nLista vacía")
            return

        ptr = self.head
        index = 0
        found = False
        positions = []

        # Recorrido circular
        while True:
            if ptr.data == item:
                positions.append(str(index))
                found = True
            
            ptr = ptr.next
            index += 1
            
            if ptr == self.head:
                break

        if found:
            print(f"Elemento {item} encontrado en la(s) ubicación(es) (índice 0-basado): {' '.join(positions)}")
        else:
            print(f"Elemento {item} no encontrado")

    def display(self):
        """
        Muestra todos los elementos de la lista desde head.
        """
        if self.is_empty():
            print("Lista vacía. Nada que imprimir.")
        else:
            print("\nImprimiendo valores (desde HEAD): ")
            ptr = self.head
            
            output = ""
            while True:
                output += str(ptr.data)
                
                # Mostrar la conexión
                if ptr.next != self.head:
                    output += " <-> "
                else:
                    # Mostrar la conexión circular
                    output += " <-> (Circular) <-> "

                ptr = ptr.next
                
                if ptr == self.head:
                    break

            # Imprimir el head de nuevo para cerrar el círculo visualmente
            output += f"[... {self.head.data} (HEAD) ...]"
            print(output)

# Función principal para el menú de interacción
def main():
    list_instance = DoublyCircularLinkedList()
    choice = 0
    
    while choice != 9:
        try:
            print("\n\n Menú principal (Lista Doblemente Enlazada Circular ) ")
            print("\nElige una opción de la siguiente lista ...\n")
            print("------------------------------------------")
            
            print("1. Insertar al principio (nuevo head)")
            print("2. Insertar al final")
            print("3. Insertar en posición (índice 0-basado, *después* del nodo en el índice)")
            print("4. Eliminar del principio (head)")
            print("5. Eliminar desde el final")
            print("6. Eliminar nodo por posición (índice 0-basado)")
            print("7. Buscar un elemento")
            print("8. Mostrar (recorrido circular)")
            print("9. Salir")
            print("------------------------------------------")
            
            user_input = input("\nIngrese su opción? ")
            choice = int(user_input)

            if choice == 9:
                print("Saliendo del programa.")
                break

            if choice not in range(1, 9):
                print("Introduzca una opción válida.")
                continue

            item = None
            loc = None

            if choice in [1, 2, 3, 7]: # Requieren un valor (item)
                try:
                    item_input = input("\nIngrese el valor del elemento: ")
                    item = int(item_input)
                except ValueError:
                    print("Entrada de valor inválida. Intente de nuevo.")
                    continue

            if choice in [3, 6]: # Requieren una posición (loc)
                try:
                    loc_input = input("Introduce la ubicación (índice 0-basado): ")
                    loc = int(loc_input)
                except ValueError:
                    print("Entrada de ubicación inválida. Intente de nuevo.")
                    continue

            if choice == 1:
                list_instance.begin_insert(item)
            elif choice == 2:
                list_instance.last_insert(item)
            elif choice == 3:
                # El valor 'item' ya se leyó si se llegó aquí
                list_instance.random_insert(item, loc)
            elif choice == 4:
                list_instance.begin_delete()
            elif choice == 5:
                list_instance.last_delete()
            elif choice == 6:
                list_instance.random_delete(loc)
            elif choice == 7:
                list_instance.search(item)
            elif choice == 8:
                list_instance.display()

        except ValueError:
            print("Opción inválida. Introduce un número del 1 al 9.")
        except Exception as e:
            print(f"\nOcurrió un error inesperado: {e}")

if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        print("\nPrograma terminado por el usuario.")
        sys.exit(0)